Accessing a management endpoint after a live reload leads to an exception (when Hibernate Search Standalone mapper is used)

### Describe the bug

Accessing a management endpoint after a live reload leads to an exception. This issue occurs with https://github.com/quarkusio/search.quarkus.io, it was discovered while working on https://github.com/quarkusio/search.quarkus.io/pull/286.

### Expected behavior

After an app is reloaded -- management endpoints can be accessed without any issues.

### Actual behavior

An exception occurs:

```java
java.lang.IllegalStateException: Failed to create indexes: Error creating synthetic bean [A3Pwpzj6LZ1U3ryq007Z-Hp-28Q]: java.lang.NullPointerException: Cannot invoke "java.util.Map.size()" because "m" is null
	at io.quarkus.search.app.indexing.IndexingService.createIndexes(IndexingService.java:213)
	at io.quarkus.search.app.indexing.IndexingService.reindex(IndexingService.java:170)
	at io.quarkus.search.app.indexing.IndexingService_Subclass.reindex$$superforward(Unknown Source)
	at io.quarkus.search.app.indexing.IndexingService_Subclass$$function$$1.apply(Unknown Source)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:73)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:62)
	at io.quarkus.arc.impl.ActivateRequestContextInterceptor.invoke(ActivateRequestContextInterceptor.java:129)
	at io.quarkus.arc.impl.ActivateRequestContextInterceptor.aroundInvoke(ActivateRequestContextInterceptor.java:33)
	at io.quarkus.arc.impl.ActivateRequestContextInterceptor_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:30)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:27)
	at io.quarkus.search.app.indexing.IndexingService_Subclass.reindex(Unknown Source)
	at io.quarkus.search.app.indexing.IndexingService.lambda$registerManagementRoutes$0(IndexingService.java:57)
	at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:191)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:279)
	at io.vertx.core.impl.ContextImpl.lambda$internalExecuteBlocking$2(ContextImpl.java:210)
	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1495)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: jakarta.enterprise.inject.CreationException: Error creating synthetic bean [A3Pwpzj6LZ1U3ryq007Z-Hp-28Q]: java.lang.NullPointerException: Cannot invoke "java.util.Map.size()" because "m" is null
	at org.hibernate.search.mapper.pojo.standalone.mapping.SearchMapping_A3Pwpzj6LZ1U3ryq007Z-Hp-28Q_Synthetic_Bean.doCreate(Unknown Source)
	at org.hibernate.search.mapper.pojo.standalone.mapping.SearchMapping_A3Pwpzj6LZ1U3ryq007Z-Hp-28Q_Synthetic_Bean.create(Unknown Source)
	at org.hibernate.search.mapper.pojo.standalone.mapping.SearchMapping_A3Pwpzj6LZ1U3ryq007Z-Hp-28Q_Synthetic_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:119)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:38)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)
	at io.quarkus.arc.generator.Default_jakarta_enterprise_context_ApplicationScoped_ContextInstances.c29(Unknown Source)
	at io.quarkus.arc.generator.Default_jakarta_enterprise_context_ApplicationScoped_ContextInstances.computeIfAbsent(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:35)
	at io.quarkus.arc.impl.ClientProxies.getApplicationScopedDelegate(ClientProxies.java:21)
	at org.hibernate.search.mapper.pojo.standalone.mapping.SearchMapping_A3Pwpzj6LZ1U3ryq007Z-Hp-28Q_Synthetic_ClientProxy.arc$delegate(Unknown Source)
	at org.hibernate.search.mapper.pojo.standalone.mapping.SearchMapping_A3Pwpzj6LZ1U3ryq007Z-Hp-28Q_Synthetic_ClientProxy.scope(Unknown Source)
	at io.quarkus.search.app.indexing.IndexingService.createIndexes(IndexingService.java:191)
	... 26 more
	Suppressed: jakarta.enterprise.inject.CreationException: Error creating synthetic bean [A3Pwpzj6LZ1U3ryq007Z-Hp-28Q]: java.lang.NullPointerException: Cannot invoke "java.util.Map.size()" because "m" is null
		at org.hibernate.search.mapper.pojo.standalone.mapping.SearchMapping_A3Pwpzj6LZ1U3ryq007Z-Hp-28Q_Synthetic_Bean.doCreate(Unknown Source)
		at org.hibernate.search.mapper.pojo.standalone.mapping.SearchMapping_A3Pwpzj6LZ1U3ryq007Z-Hp-28Q_Synthetic_Bean.create(Unknown Source)
		at org.hibernate.search.mapper.pojo.standalone.mapping.SearchMapping_A3Pwpzj6LZ1U3ryq007Z-Hp-28Q_Synthetic_Bean.create(Unknown Source)
		at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:119)
		at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:38)
		at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)
		at io.quarkus.arc.generator.Default_jakarta_enterprise_context_ApplicationScoped_ContextInstances.c29(Unknown Source)
		at io.quarkus.arc.generator.Default_jakarta_enterprise_context_ApplicationScoped_ContextInstances.computeIfAbsent(Unknown Source)
		at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:35)
		at io.quarkus.arc.impl.ClientProxies.getApplicationScopedDelegate(ClientProxies.java:21)
		at org.hibernate.search.mapper.pojo.standalone.mapping.SearchMapping_A3Pwpzj6LZ1U3ryq007Z-Hp-28Q_Synthetic_ClientProxy.arc$delegate(Unknown Source)
		at org.hibernate.search.mapper.pojo.standalone.mapping.SearchMapping_A3Pwpzj6LZ1U3ryq007Z-Hp-28Q_Synthetic_ClientProxy.backend(Unknown Source)
		at io.quarkus.search.app.indexing.Rollover.client(Rollover.java:105)
		at io.quarkus.search.app.indexing.Rollover.recoverInconsistentAliases(Rollover.java:92)
		at io.quarkus.search.app.indexing.IndexingService.createIndexes(IndexingService.java:205)
		... 26 more
	Caused by: java.lang.NullPointerException: Cannot invoke "java.util.Map.size()" because "m" is null
		at java.base/java.util.HashMap.putMapEntries(HashMap.java:495)
		at java.base/java.util.LinkedHashMap.<init>(LinkedHashMap.java:385)
		at io.quarkus.hibernate.search.standalone.elasticsearch.runtime.HibernateSearchStandaloneRecorder$1.apply(HibernateSearchStandaloneRecorder.java:110)
		at io.quarkus.hibernate.search.standalone.elasticsearch.runtime.HibernateSearchStandaloneRecorder$1.apply(HibernateSearchStandaloneRecorder.java:103)
		at org.hibernate.search.mapper.pojo.standalone.mapping.SearchMapping_A3Pwpzj6LZ1U3ryq007Z-Hp-28Q_Synthetic_Bean.createSynthetic(Unknown Source)
		... 41 more
Caused by: java.lang.NullPointerException: Cannot invoke "java.util.Map.size()" because "m" is null
	at java.base/java.util.HashMap.putMapEntries(HashMap.java:495)
	at java.base/java.util.LinkedHashMap.<init>(LinkedHashMap.java:385)
	at io.quarkus.hibernate.search.standalone.elasticsearch.runtime.HibernateSearchStandaloneRecorder$1.apply(HibernateSearchStandaloneRecorder.java:110)
	at io.quarkus.hibernate.search.standalone.elasticsearch.runtime.HibernateSearchStandaloneRecorder$1.apply(HibernateSearchStandaloneRecorder.java:103)
	at org.hibernate.search.mapper.pojo.standalone.mapping.SearchMapping_A3Pwpzj6LZ1U3ryq007Z-Hp-28Q_Synthetic_Bean.createSynthetic(Unknown Source)
	... 39 more
```

### How to Reproduce?

1. Run a Quarkus app in dev mode
2. Make some changes and save them so that the app is reloaded
3. After the app is reloaded, trying to access a management endpoint results in an exception.

### Output of `uname -a` or `ver`

Linux fedora 6.8.11-200.fc39.x86_64 #1 SMP PREEMPT_DYNAMIC Sun May 26 20:05:41 UTC 2024 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.10" 2024-01-16 OpenJDK Runtime Environment Temurin-17.0.10+7 (build 17.0.10+7) OpenJDK 64-Bit Server VM Temurin-17.0.10+7 (build 17.0.10+7, mixed mode, sharing)

### Quarkus version or git rev

3.11.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.3 (21122926829f1ead511c958d89bd2f672198ae9f)

### Additional information

I'm opening the issue as is so that we don't forget to investigate it. If someone else picks it up and has questions about reproducing it or running search.quarkus.io, feel free to ping me.

--- a/extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/VertxHttpProcessor.java
+++ b/extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/VertxHttpProcessor.java
@@ -265,16 +265,14 @@ class VertxHttpProcessor {
         List<RouteBuildItem> redirectRoutes = new ArrayList<>();
         boolean frameworkRouterCreated = false;
         boolean mainRouterCreated = false;
-        boolean managementRouterCreated = false;
 
         boolean isManagementInterfaceEnabled = managementBuildTimeConfig.enabled;
+        if (isManagementInterfaceEnabled) {
+            managementRouter = recorder.initializeRouter(vertx.getVertx());
+        }
 
         for (RouteBuildItem route : routes) {
             if (route.isManagement() && isManagementInterfaceEnabled) {
-                if (!managementRouterCreated) {
-                    managementRouter = recorder.initializeRouter(vertx.getVertx());
-                    managementRouterCreated = true;
-                }
                 recorder.addRoute(managementRouter, route.getRouteFunction(), route.getHandler(), route.getType());
             } else if (nonApplicationRootPath.isDedicatedRouterRequired() && route.isRouterFramework()) {
                 // Non-application endpoints on a separate path

--- a/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/devmode/management/LiveReloadManagementBean.java
+++ b/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/devmode/management/LiveReloadManagementBean.java
@@ -1,0 +1,16 @@
+package io.quarkus.vertx.http.devmode.management;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Named;
+
+@Named
+@ApplicationScoped
+public class LiveReloadManagementBean {
+
+    private final String value = "string1";
+
+    String string() {
+        return value;
+    }
+
+}

--- a/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/devmode/management/LiveReloadManagementEndpoint.java
+++ b/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/devmode/management/LiveReloadManagementEndpoint.java
@@ -1,0 +1,35 @@
+package io.quarkus.vertx.http.devmode.management;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.enterprise.event.Observes;
+
+import io.quarkus.vertx.http.ManagementInterface;
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.client.HttpRequest;
+import io.vertx.ext.web.client.WebClient;
+
+// Case 3: import jakarta.inject.Inject;
+
+@ApplicationScoped
+public class LiveReloadManagementEndpoint {
+
+    // Case 3: @Inject
+    // Case 3: LiveReloadManagementHandlerAsCDIBean handler;
+
+    void managementRoutes(@Observes ManagementInterface mi) {
+        Router router = mi.router();
+        // Case 1: testing name change:
+        router.route("/manage-1")
+                .produces("text/plain")
+                .handler(rc -> rc.response().end(WebClient.class.hashCode() + "-" + HttpRequest.class.hashCode()));
+
+        // Test that a new handler using some CDI bean can be loaded:
+        // Case 2: router.route("/manage-cdi")
+        // Case 2:         .produces("text/plain")
+        // Case 2:         .handler(new LiveReloadManagementHandler());
+
+        // Case 3: router.route("/manage-bean-handler")
+        // Case 3:         .produces("text/plain")
+        // Case 3:         .handler(handler);
+    }
+}

--- a/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/devmode/management/LiveReloadManagementHandler.java
+++ b/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/devmode/management/LiveReloadManagementHandler.java
@@ -1,0 +1,14 @@
+package io.quarkus.vertx.http.devmode.management;
+
+import io.quarkus.arc.Arc;
+import io.vertx.core.Handler;
+import io.vertx.ext.web.RoutingContext;
+
+public class LiveReloadManagementHandler implements Handler<RoutingContext> {
+
+    @Override
+    public void handle(RoutingContext event) {
+        LiveReloadManagementBean managementBean = Arc.container().instance(LiveReloadManagementBean.class).get();
+        event.response().end(managementBean.string());
+    }
+}

--- a/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/devmode/management/LiveReloadManagementHandlerAsCDIBean.java
+++ b/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/devmode/management/LiveReloadManagementHandlerAsCDIBean.java
@@ -1,0 +1,17 @@
+package io.quarkus.vertx.http.devmode.management;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Named;
+
+import io.vertx.core.Handler;
+import io.vertx.ext.web.RoutingContext;
+
+@Named
+@ApplicationScoped
+public class LiveReloadManagementHandlerAsCDIBean implements Handler<RoutingContext> {
+
+    @Override
+    public void handle(RoutingContext event) {
+        event.response().end("I'm a CDI bean handler.");
+    }
+}

--- a/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/devmode/management/LiveReloadManagementTest.java
+++ b/extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/devmode/management/LiveReloadManagementTest.java
@@ -1,0 +1,107 @@
+package io.quarkus.vertx.http.devmode.management;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.concurrent.TimeUnit;
+
+import org.awaitility.Awaitility;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusDevModeTest;
+import io.restassured.RestAssured;
+
+public class LiveReloadManagementTest {
+
+    private static final String APP_PROPS = """
+            quarkus.class-loading.reloadable-artifacts=io.vertx:vertx-web-client
+            quarkus.management.enabled=true
+            """;
+
+    @RegisterExtension
+    static final QuarkusDevModeTest test = new QuarkusDevModeTest()
+            .withApplicationRoot((jar) -> jar
+                    .addAsResource(new StringAsset(APP_PROPS), "application.properties")
+                    .addClasses(LiveReloadManagementEndpoint.class));
+
+    private final String managementPath = "http://localhost:9000";
+
+    @Test
+    public void test() {
+        // 1. We start with a single management endpoint /manage-1
+        String firstClassToString = RestAssured.get(managementPath + "/manage-1")
+                .then()
+                .statusCode(200)
+                .extract().body().asString();
+        // 2. We change the path of that endpoint, esentailly testing adding new and removing old one:
+        test.modifySourceFile(LiveReloadManagementEndpoint.class, s -> s.replace("\"/manage-1\"", "\"/manage-2\""));
+
+        // 2.1 Wait for the reload to be done, and first make sure that the "new" endpoint is there:
+        Awaitility.await()
+                .atMost(5, TimeUnit.SECONDS)
+                .untilAsserted(() -> {
+                    String secondClassToString = RestAssured.get(managementPath + "/manage-2")
+                            .then()
+                            .statusCode(200)
+                            .extract().body().asString();
+                    assertThat(firstClassToString).isNotEqualTo(secondClassToString);
+                });
+
+        // 2.2 Since the above check was a success, it means that the app was reloaded and we are sure that
+        // the /manage-1 is no longer available:
+        RestAssured.get(managementPath + "/manage-1")
+                .then()
+                .statusCode(404);
+
+        // 3. Now we want to add another endpoint that attempts to access a CDI bean
+        // 4. First let's make sure that it is not there already:
+        RestAssured.get(managementPath + "/manage-cdi").then().statusCode(404);
+
+        // 5. Add/update all necessary source files:
+        test.addSourceFile(LiveReloadManagementBean.class);
+        test.addSourceFile(LiveReloadManagementHandler.class);
+        test.modifySourceFile(LiveReloadManagementEndpoint.class, s -> s.replace("// Case 2: ", ""));
+
+        // 6. Now wait for the app to get reloaded:
+        Awaitility.await()
+                .atMost(5, TimeUnit.SECONDS)
+                .untilAsserted(() -> {
+                    String response = RestAssured.get(managementPath + "/manage-cdi")
+                            .then()
+                            .statusCode(200)
+                            .extract().body().asString();
+                    assertThat(response).isEqualTo("string1");
+                });
+
+        // 7. Update the bean used in the handler:
+        test.modifySourceFile(LiveReloadManagementBean.class, s -> s.replace("string1", "string2"));
+
+        // 8. Wait for the app to get reloaded and management endpoint should be using the reloaded CDI bean:
+        Awaitility.await()
+                .atMost(5, TimeUnit.SECONDS)
+                .untilAsserted(() -> {
+                    String response = RestAssured.get(managementPath + "/manage-cdi")
+                            .then()
+                            .statusCode(200)
+                            .extract().body().asString();
+                    assertThat(response).isEqualTo("string2");
+                });
+
+        // 9. Now let's add yet another handler, this time the handler is injected as a bean. (case 3)
+        test.addSourceFile(LiveReloadManagementHandlerAsCDIBean.class);
+        test.modifySourceFile(LiveReloadManagementEndpoint.class, s -> s.replace("// Case 3: ", ""));
+
+        // 10. Now wait for the app to get reloaded:
+        Awaitility.await()
+                .atMost(5, TimeUnit.SECONDS)
+                .untilAsserted(() -> {
+                    String response = RestAssured.get(managementPath + "/manage-bean-handler")
+                            .then()
+                            .statusCode(200)
+                            .extract().body().asString();
+                    assertThat(response).isEqualTo("I'm a CDI bean handler.");
+                });
+    }
+
+}

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
@@ -526,6 +526,7 @@ public class VertxHttpRecorder {
         if (managementRouter != null && managementRouter.getValue() != null) {
             // Add body handler and cors handler
             var mr = managementRouter.getValue();
+            boolean hasManagementRoutes = !mr.getRoutes().isEmpty();
 
             addHotReplacementHandlerIfNeeded(mr);
 
@@ -548,16 +549,25 @@ public class VertxHttpRecorder {
             Handler<HttpServerRequest> handler = HttpServerCommonHandlers.enforceDuplicatedContext(mr);
             handler = HttpServerCommonHandlers.applyProxy(managementConfiguration.getValue().proxy, handler, vertx);
 
-            event.select(ManagementInterface.class).fire(new ManagementInterfaceImpl(managementRouter.getValue()));
+            int routesBeforeMiEvent = mr.getRoutes().size();
+            event.select(ManagementInterface.class).fire(new ManagementInterfaceImpl(mr));
 
-            VertxHttpRecorder.managementRouterDelegate = handler;
-            if (VertxHttpRecorder.managementRouter == null) {
-                VertxHttpRecorder.managementRouter = new Handler<HttpServerRequest>() {
-                    @Override
-                    public void handle(HttpServerRequest event) {
-                        VertxHttpRecorder.managementRouterDelegate.handle(event);
-                    }
-                };
+            // It may be that no build steps produced any management routes.
+            // But we still want to give a chance to the "ManagementInterface event" to collect any
+            // routes that users may have provided through observing this event.
+            //
+            // Hence, we only initialize the `managementRouter` router when we either had some routes from extensions (`hasManagementRoutes`)
+            // or if the event collected some routes (`routesBeforeMiEvent < routesAfterMiEvent`)
+            if (hasManagementRoutes || routesBeforeMiEvent < mr.getRoutes().size()) {
+                VertxHttpRecorder.managementRouterDelegate = handler;
+                if (VertxHttpRecorder.managementRouter == null) {
+                    VertxHttpRecorder.managementRouter = new Handler<HttpServerRequest>() {
+                        @Override
+                        public void handle(HttpServerRequest event) {
+                            VertxHttpRecorder.managementRouterDelegate.handle(event);
+                        }
+                    };
+                }
             }
         }
     }
@@ -1577,7 +1587,7 @@ public class VertxHttpRecorder {
 
         @Override
         public void handle(RoutingContext event) {
-            Thread.currentThread().setContextClassLoader( currentCl );
+            Thread.currentThread().setContextClassLoader(currentCl);
             hotReplacementHandler.handle(event);
         }
     }

--- a/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
+++ b/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java
@@ -204,6 +204,7 @@ public class VertxHttpRecorder {
 
     final RuntimeValue<ManagementInterfaceConfiguration> managementConfiguration;
     private static volatile Handler<HttpServerRequest> managementRouter;
+    private static volatile Handler<HttpServerRequest> managementRouterDelegate;
 
     public VertxHttpRecorder(HttpBuildTimeConfig httpBuildTimeConfig,
             ManagementInterfaceBuildTimeConfig managementBuildTimeConfig,
@@ -437,34 +438,14 @@ public class VertxHttpRecorder {
 
         Handler<HttpServerRequest> root;
         if (rootPath.equals("/")) {
-            if (hotReplacementHandler != null) {
-                //recorders are always executed in the current CL
-                ClassLoader currentCl = Thread.currentThread().getContextClassLoader();
-                httpRouteRouter.route().order(RouteConstants.ROUTE_ORDER_HOT_REPLACEMENT)
-                        .handler(new Handler<RoutingContext>() {
-                            @Override
-                            public void handle(RoutingContext event) {
-                                Thread.currentThread().setContextClassLoader(currentCl);
-                                hotReplacementHandler.handle(event);
-                            }
-                        });
-            }
+            addHotReplacementHandlerIfNeeded(httpRouteRouter);
             root = httpRouteRouter;
         } else {
             Router mainRouter = mainRouterRuntimeValue.isPresent() ? mainRouterRuntimeValue.get().getValue()
                     : Router.router(vertx.get());
             mainRouter.mountSubRouter(rootPath, httpRouteRouter);
 
-            if (hotReplacementHandler != null) {
-                ClassLoader currentCl = Thread.currentThread().getContextClassLoader();
-                mainRouter.route().order(RouteConstants.ROUTE_ORDER_HOT_REPLACEMENT).handler(new Handler<RoutingContext>() {
-                    @Override
-                    public void handle(RoutingContext event) {
-                        Thread.currentThread().setContextClassLoader(currentCl);
-                        hotReplacementHandler.handle(event);
-                    }
-                });
-            }
+            addHotReplacementHandlerIfNeeded(mainRouter);
             root = mainRouter;
         }
 
@@ -546,6 +527,8 @@ public class VertxHttpRecorder {
             // Add body handler and cors handler
             var mr = managementRouter.getValue();
 
+            addHotReplacementHandlerIfNeeded(mr);
+
             mr.route().last().failureHandler(
                     new QuarkusErrorHandler(launchMode.isDevOrTest(), httpConfiguration.unhandledErrorContentTypeDefault));
 
@@ -567,7 +550,24 @@ public class VertxHttpRecorder {
 
             event.select(ManagementInterface.class).fire(new ManagementInterfaceImpl(managementRouter.getValue()));
 
-            VertxHttpRecorder.managementRouter = handler;
+            VertxHttpRecorder.managementRouterDelegate = handler;
+            if (VertxHttpRecorder.managementRouter == null) {
+                VertxHttpRecorder.managementRouter = new Handler<HttpServerRequest>() {
+                    @Override
+                    public void handle(HttpServerRequest event) {
+                        VertxHttpRecorder.managementRouterDelegate.handle(event);
+                    }
+                };
+            }
+        }
+    }
+
+    private void addHotReplacementHandlerIfNeeded(Router router) {
+        if (hotReplacementHandler != null) {
+            //recorders are always executed in the current CL
+            ClassLoader currentCl = Thread.currentThread().getContextClassLoader();
+            router.route().order(RouteConstants.ROUTE_ORDER_HOT_REPLACEMENT)
+                    .handler(new HotReplacementRoutingContextHandler(currentCl));
         }
     }
 
@@ -1565,6 +1565,20 @@ public class VertxHttpRecorder {
         @Override
         public boolean getAsBoolean() {
             return true;
+        }
+    }
+
+    private static class HotReplacementRoutingContextHandler implements Handler<RoutingContext> {
+        private final ClassLoader currentCl;
+
+        public HotReplacementRoutingContextHandler(ClassLoader currentCl) {
+            this.currentCl = currentCl;
+        }
+
+        @Override
+        public void handle(RoutingContext event) {
+            Thread.currentThread().setContextClassLoader( currentCl );
+            hotReplacementHandler.handle(event);
         }
     }
 }

